tests/gpflow/experimental/check_shapes/test_check_shapes.py

```diff
@@ -382,33 +394,65 @@ def h(a: TensorType, b: TensorType) -> TensorType:
             "b: [d...]",
             "return: [d...]",
         )
-        def f(a: TensorType, b: TensorType) -> TensorType:
+        def f(a: TestShaped, b: TestShaped) -> TestShaped:
             return a
 
         f(t(2, 3), t(2, 4))  # Wrong shape, but checks disabled.
 
     f(t(2, 3), t(2, 4))  # Wrong shape, but checks were disable when function was created.
-    get_check_shapes(f)(h)  # Don't crash.
+    get_check_shapes(f)(h)  # pylint: disable=not-callable  # Don't crash.
 
     @check_shapes(
         "a: [d...]",
         "b: [d...]",
         "return: [d...]",
     )
-    def g(a: TensorType, b: TensorType) -> TensorType:
+    def g(a: TestShaped, b: TestShaped) -> TestShaped:
         return a
 
     with pytest.raises(ShapeMismatchError):
         g(t(2, 3), t(2, 4))
-        get_check_shapes(g)(h)  # Don't crash.
+        get_check_shapes(g)(h)  # pylint: disable=not-callable    # Don't crash.
 
     with disable_check_shapes():
         g(t(2, 3), t(2, 4))  # Wrong shape, but checks disabled.
-        get_check_shapes(g)(h)  # Don't crash.
+        get_check_shapes(g)(h)  # pylint: disable=not-callable    # Don't crash.
 
     with pytest.raises(ShapeMismatchError):
         g(t(2, 3), t(2, 4))
-        get_check_shapes(g)(h)  # Don't crash.
+        get_check_shapes(g)(h)  # pylint: disable=not-callable    # Don't crash.
+
+
+def test_check_shapes__error_message() -> None:
+    # Here we're just testing that error message formatting is wired together sanely. For more
+    # thorough tests of error formatting, see test_errors.py
+
+    @check_shapes(
+        "a: [d1, d2]",
+        "b: [d1, d3]",
+        "return: [d2, d3]",
+    )
+    def f(a: TestShaped, b: TestShaped) -> TestShaped:
+        return t(3, 4)
+
+    with pytest.raises(ShapeMismatchError) as e:
+        f(t(2, 3), t(3, 4))
+
+    (message,) = e.value.args
+    assert (
+        f"""
+Tensor shape mismatch in call to function.
+  Function: test_check_shapes__error_message.<locals>.f
+    Declared: {__file__}:430
```

Comment:
Would be nice to avoid having to hardcode the line number so the test doesn't break whenever anyone updated the tests above it. Can you get the check shape line number using something like https://stackoverflow.com/questions/3056048/filename-and-line-number-of-python-script ?

Reply:
> Would be nice to avoid having to hardcode the line number so the test doesn't break whenever anyone updated the tests above it. Can you get the check shape line number using something like https://stackoverflow.com/questions/3056048/filename-and-line-number-of-python-script ?

It would be nice, but I don't like the idea of using `inspect` and looking a the stack, because that's exactly what we're doing in the production code. I have seen a couple of instances of people writing the same algorithm in the production code and the test and then checking that they give the same result. That doesn't work if the algorithm is fundamentally wrong - you just check that the wrong result is the same as the same wrong result.